web/src/message_events_util.ts

```diff
@@ -22,35 +41,33 @@ export function maybe_add_narrowed_messages(messages, msg_list, callback, attemp
             narrow: JSON.stringify(narrow_state.public_search_terms()),
         },
         timeout: 5000,
-        success(data) {
+        success(raw_data) {
             if (!narrow_state.is_message_feed_visible() || msg_list !== message_lists.current) {
                 // We unnarrowed or moved to Recent Conversations in the meantime.
                 return;
             }
+            
+            const data = msg_match_narrow_api_response_schema.parse(raw_data);
 
-            let new_messages = [];
-            const elsewhere_messages = [];
+            let new_messages:Message[] = [];
+            const elsewhere_messages:Message[] = [];
 
             for (const elem of messages) {
                 if (Object.hasOwn(data.messages, elem.id)) {
-                    util.set_match_data(elem, data.messages[elem.id]);
+                    util.set_match_data(elem, data.messages[elem.id]!);
                     new_messages.push(elem);
                 } else {
                     elsewhere_messages.push(elem);
                 }
             }
 
-            // This second call to process_new_message in the
-            // insert_new_messages code path is designed to replace
-            // our slightly stale message object with the latest copy
-            // from the message_store. This helps in very rare race
+            // We replace our slightly stale message object with the latest
+            // copy from the message_store. This helps in very rare race
             // conditions, where e.g. the current user's name was
             // edited in between when they sent the message and when
             // we hear back from the server and can echo the new
             // message.
-            new_messages = new_messages.map((message) =>
-                message_helper.process_new_message(message),
```

Comment:
We are doing something which is pretty evil to TypeScript. `message_helper.process_new_message` takes `RawMessage` and throws out `Message`(as the name suggests). But here we are sending it a `Message` type which differs **a lot** from the RawMessage type. I have tried to shove the problem under the carpet for now by passing `message_helper.process_new_message` but we would have to find a solution -- and extending the type of `process_new_message` is not one IMO, since that will dilutes the type structure of that function.

Reply:
Creating another function something like `reprocess_msg` could be one, but creating one only to handle this very very rare usecase -- I need some advice on this.

Reply:
So I tried to dig a bit into this and I think the only reason why we are calling this function is to get the `cached_message` (that's what the first conditional of `process_new_message` suggests). I think writing a new function is worth it or maybe just writing the logic here, to get the cached message. 
What do you think so?

If we use this approach of passing the function from `echo.js` for now it will be a hindrance when you try to migrate `echo.js`

Reply:
Hmm, your point makes sense according to my understanding -- new version has the discussed change.

Reply:
Can we do it in a prep commit?

Reply:
This prep commit makes it very clear, thanks!